src/extensions/rppgrpc/rppgrpc/client_reactor.hpp

```diff
@@ -0,0 +1,303 @@
+            }
+            Destroy();
+        }
+
+    private:
+        void Destroy()
+        {
+            m_disposable.dispose();
+            delete this;
+        }
+
+    private:
+        Observer                m_observer;
+        rpp::disposable_wrapper m_disposable;
+
+        Response m_read{};
+
+        std::mutex         m_write_mutex{};
+        std::list<Request> m_write{};
+    };
+
+    template<rpp::constraint::observer Observer>
+    class client_read_reactor final : public grpc::ClientReadReactor<rpp::utils::extract_observer_type_t<Observer>>
+    {
+        using Response = rpp::utils::extract_observer_type_t<Observer>;
+        using Base     = grpc::ClientReadReactor<Response>;
+
+    public:
+        template<rpp::constraint::decayed_same_as<Observer> TObserver>
+            requires (!rpp::constraint::decayed_same_as<TObserver, client_read_reactor<Observer>>)
+        explicit client_read_reactor(TObserver&& events)
+            : m_observer{std::forward<TObserver>(events)}
+        {
+        }
+
+        client_read_reactor(client_read_reactor&&) = delete;
+
+        void Init()
+        {
+            Base::StartCall();
+            Base::StartRead(&m_read);
+        }
+
+    private:
+        void OnReadDone(bool ok) override
+        {
+            if (!ok)
+            {
+                m_observer.on_error(std::make_exception_ptr(rppgrpc::utils::reactor_faield{"OnReadDone is not ok"}));
+                Destroy();
+                return;
+            }
+            m_observer.on_next(m_read);
+            Base::StartRead(&m_read);
+        }
+
+        void OnDone(const grpc::Status& s) override
+        {
+            if (s.ok())
+            {
+                m_observer.on_completed();
+            }
+            else
+            {
+                m_observer.on_error(std::make_exception_ptr(rppgrpc::utils::reactor_faield{s.error_message()}));
+            }
+            Destroy();
+        }
+
+    private:
+        void Destroy()
+        {
+            delete this;
+        }
+
+    private:
+        Observer m_observer;
+        Response m_read{};
+    };
+} // namespace rppgrpc::details
+namespace rppgrpc
+{
+    template<typename AsyncInMethod,
+             std::derived_from<AsyncInMethod> Async,
+             rpp::constraint::observable      Observable,
+             rpp::constraint::observer        Observer>
+    void add_client_reactor(member_bidi_function_ptr<AsyncInMethod, Observable, Observer> method,
+                            Async&                                                        async,
+                            grpc::ClientContext*                                          context,
+                            const Observable&                                             requests,
+                            Observer&&                                                    responses)
```

Comment:
Is it possible to add a `add_client_reactor` specialization to directly return a responses observable? Instead of client having to create a subject manually everytime?

Reply:
it is possible, but actually it means, that observable would be hot in this case and it can start to emit values till returning from such an function. This signature allows to avoid id. Also in this case user can select where to use serialized/replay/behavior subjects or not. In case of direct returning RPP should choose by itself

Reply:
Is it possible for a user to get a cold observable out of it instead? Subscription would then trigger the connection

Reply:
Theoretically - yes, but what if user firstly want to emit values and then subscribe to observable later?

Reply:
Agree that is one possibility but maybe having both possibilities would help as it seems a bit tedious to always have to work with hot observables

Reply:
Do you mean, any of this operations should start connection?

Reply:
It is already starting a connection immediately here no? Was wondering if the connection could be driven by suscription instead directly but of course it is always possible why your approach by wrapping it behind a `defer` I guess